
Q1. What is multithreading in python? Why is it used? Name the module used to handle threads in python. Why threading module used?

Multithreading in Python is the ability to run multiple threads simultaneously. This can be useful for tasks that are computationally expensive, as it can allow the program to continue to be responsive while the tasks are running.

The threading module is used to handle threads in Python. It provides a number of functions for creating, managing, and synchronizing threads.

The threading module is useful because it provides a simple and easy-to-use interface for working with threads. It also provides a number of features that make it easy to write safe and efficient multithreaded code.

Q2. Explain the following functions:

activeCount(): Returns the number of active threads.

currentThread(): Returns the current thread.

enumerate(): Returns a list of all active threads.

Q3. Explain the following functions:

run(): The run() method is the method that is executed when a thread is started.

start(): The start() method starts a thread.

join(): The join() method blocks the current thread until the target thread has finished executing.

isAlive(): The isAlive() method returns True if the thread is still alive, and False otherwise.

In [1]:
#Q4. Write a python program to create two threads. Thread one must print the list of squares and thread two must print the list of cubes.

import threading

def print_squares(n):
  for i in range(n + 1):
    print(i * i)

def print_cubes(n):
  for i in range(n + 1):
    print(i * i * i)

if __name__ == "__main__":
  t1 = threading.Thread(target=print_squares, args=(10,))
  t2 = threading.Thread(target=print_cubes, args=(10,))

  t1.start()
  t2.start()

  t1.join()
  t2.join()


0
1
4
9
16
25
36
49
64
81
100
0
1
8
27
64
125
216
343
512
729
1000



Q1. What is multithreading in python? Why is it used? Name the module used to handle threads in python. Why threading module used?

Multithreading in Python is the ability to run multiple threads simultaneously. This can be useful for tasks that are computationally expensive, as it can allow the program to continue to be responsive while the tasks are running.

The threading module is used to handle threads in Python. It provides a number of functions for creating, managing, and synchronizing threads.

The threading module is useful because it provides a simple and easy-to-use interface for working with threads. It also provides a number of features that make it easy to write safe and efficient multithreaded code.

Q2. Explain the following functions:

activeCount(): Returns the number of active threads.
currentThread(): Returns the current thread.
enumerate(): Returns a list of all active threads.
Q3. Explain the following functions:

run(): The run() method is the method that is executed when a thread is started.
start(): The start() method starts a thread.
join(): The join() method blocks the current thread until the target thread has finished executing.
isAlive(): The isAlive() method returns True if the thread is still alive, and False otherwise.
Q4. Write a python program to create two threads. Thread one must print the list of squares and thread two must print the list of cubes.

Python
import threading

def print_squares(n):
  for i in range(n + 1):
    print(i * i)

def print_cubes(n):
  for i in range(n + 1):
    print(i * i * i)

if __name__ == "__main__":
  t1 = threading.Thread(target=print_squares, args=(10,))
  t2 = threading.Thread(target=print_cubes, args=(10,))

  t1.start()
  t2.start()

  t1.join()
  t2.join()
Use code with caution. Learn more
Output:

0
1
4
9
16
25
36
49
64
81
100
0
1
8
27
64
125
216
343
512
729
1000
Q5. State advantages and disadvantages of multithreading.

Advantages of multithreading:

Improved performance: Multithreading can improve the performance of a program by allowing it to run multiple tasks simultaneously. This is especially beneficial for computationally expensive tasks.

Increased responsiveness: Multithreading can make a program more responsive by allowing it to continue to be responsive while computationally expensive tasks are running.
Simplified code: Multithreading can simplify code by making it easier to write code that is concurrent and parallel.
Disadvantages of multithreading:

Increased complexity: Multithreading can increase the complexity of a program, as it requires the programmer to think about how to synchronize threads and avoid race conditions.
Increased debugging difficulty: Multithreaded programs can be more difficult to debug than single-threaded programs, as the programmer needs to consider the possibility of race conditions and other concurrency-related problems.
Reduced performance on some systems: Multithreading can reduce performance on some systems, such as systems with a single CPU.
Q6. Explain deadlocks and race conditions.

Deadlock: A deadlock is a situation where two or more threads are waiting for each other to finish executing, and neither thread can finish executing because it is waiting for the other thread.

Race condition: A race condition is a situation where two or more threads are accessing the same data at the same time, and the outcome of the program depends on the order in which the threads access the data.

Deadlocks and race conditions can be difficult to detect and debug, and they can lead to unpredictable and unexpected behavior in a program.